File: DEVELOP/DFT.java
import com.cycling74.max.*;

public class DFTProcessor extends MaxObject {
    // Constructor: define the number of inlets and outlets
    public DFTProcessor() {
        // One inlet to receive the signal
        declareInlet(0);
        // Two outlets: one for the real part and one for the imaginary part
        declareOutlets(new int[]{DataTypes.LIST, DataTypes.LIST});
    }

    // This method processes the list of incoming time-domain samples
    public void list(Atom[] args) {
        int N = args.length;
        double[] x = new double[N];

        // Convert input list (atoms) to a double array
        for (int i = 0; i < N; i++) {
            x[i] = args[i].toDouble();
        }

        // Arrays for real and imaginary parts of the DFT result
        double[] X_real = new double[N];
        double[] X_imag = new double[N];

        // Compute the DFT
        for (int k = 0; k < N; k++) {
            double realSum = 0;
            double imagSum = 0;
            for (int n = 0; n < N; n++) {
                double angle = (2 * Math.PI * k * n) / N;
                realSum += x[n] * Math.cos(angle);
                imagSum -= x[n] * Math.sin(angle); // Imaginary part
            }
            X_real[k] = realSum;
            X_imag[k] = imagSum;
        }

        // Output the real part
        Atom[] realOutput = new Atom[N];
        Atom[] imagOutput = new Atom[N];
        for (int i = 0; i < N; i++) {
            realOutput[i] = Atom.newAtom(X_real[i]);
            imagOutput[i] = Atom.newAtom(X_imag[i]);
        }
        // Output real part via the first outlet
        outlet(0, realOutput);
        // Output imaginary part via the second outlet
        outlet(1, imagOutput);
    }
}

File: DEVELOP/DFT.js
/*
The DFT converts a time-domain signal into its frequency-domain representation.
	N is the number of samples,
	k is the index for the frequency bin (0 to N-1),
	n is the index for the time sample (0 to N-1),
	j is the imaginary unit.
This script accepts an array of time-domain samples (via the list message), 
computes the DFT, and outputs the real and imaginary parts separately for each frequency bin.
*/

inlets = 1;  // One inlet for receiving input
outlets = 2; // Two outlets: one for real part, one for imaginary part

setinletassist(0,"(list) time-domain signal");
setoutletassist(0,"(list) real part");
setoutletassist(1,"(list) imaginary part");

// Function to compute DFT
function list() {
    var x = arrayfromargs(arguments); // Input time-domain signal (array of numbers)
    var N = x.length; // Number of samples
    var X_real = [];  // Array for real parts of the DFT result
    var X_imag = [];  // Array for imaginary parts of the DFT result

    // DFT computation
    for (var k = 0; k < N; k++) {
        var realSum = 0;
        var imagSum = 0;
        for (var n = 0; n < N; n++) {
            var angle = (2 * Math.PI * k * n) / N;
            realSum += x[n] * Math.cos(angle);
            imagSum -= x[n] * Math.sin(angle);  // Note the negative sign for the imaginary part
        }
        X_real[k] = realSum;
        X_imag[k] = imagSum;
    }

    // Output the real and imaginary parts through separate outlets
	outlet(1, X_imag);  // Output imaginary part
    outlet(0, X_real);  // Output real part
    
	//post("real " + X_real + "\n");
	//post("img " + X_imag + "\n");

}


